### PR TITLE
#29 add initial graph ql schema

### LINKED ISSUE #29: Add initial backend GraphQL schema
Based on initial Dgraph schema, add scripts/schema to upload GraphQL schema to Dgraph instance.

### PR DESCRIPTION
**Pull request by katherine-sussol**
Fixes #29 

Schema/scripts modified to allow us to work with GraphQL endpoint exposed by Dgraph.

- Types have been mapped 1-1 with existing Dgraph instance
  - This means our Dgraph instance will be backward-compatible with existing Dgraph queries (though we have the option of rewriting the V1 REST api to consume GraphQL directly instead of GraphQL+-)
  - Ratel UI can still be used to browse the underlying Dgraph structure 
- init.sh will now setup by uploading the following:
  - Schema (schema.gql) via localhost:8080/admin/schema
  - Data (insert.gql) via localhost:8080/graphql

**Sample Query** 
Submit via POST to localhost:8080/graphql:
```
query {
    queryEntity (filter: { type: { eq: "unit_of_use" } }) {
        code
        description
    }
}
```

**Sample Response**
```
{
    "data": {
        "queryEntity": [
            {
                "code": "5589F4BF",
                "description": "Paracetamol oral tablet, 300mg"
            },
            {
                "code": "368C74BE",
                "description": "Amoxicillin oral tablet, 500mg"
            },
            {
                "code": "368C74BF",
                "description": "Amoxicillin oral capsule, 250mg"
            }
        ]
    }
}
```

### CODE REVIEW

**Comment by katherine-sussol** [comment]
Wasn't sure whether or not to delete data.json and data.rdf as I assume from now we will be using graphQL endpoint to do uploads going forward?

**Comment by wlthomson** [comment]
@katherine-sussol Nice! Looks good. Will do a proper review tomorrow. 

Btw, generally we just reference the issue for all the "metadata" (e.g. labels, projects, milestones), so no need to add them to the PR :).

**Comment by katherine-sussol** [comment]
@wlthomson Cool, thanks for pointing that out, looks like I got a bit click happy each time I saw a gear icon :). Happy to swap to using 'fixes' if it's an established convention (and supported with automation). 
Initial reason for using 'resolves' was that does seem slightly odd to 'fix' something that isn't technically broken ;)

**Comment by mark-prins** [comment]
I just use `Fixes` for all PRs - consistency over sensibility. Don't mind either way, although if the automation relies on fixes then better to use that exclusively.

File: dgraph/query.sh [hunk 1]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash
 
-curl -s -H "Content-Type: application/graphql+-" "localhost:8080/query" -XPOST --data-binary "@$1" | python3 -m json.tool
\ No newline at end of file
+curl -s -H "Content-Type: application/graphql" "localhost:8080/graphql" -XPOST --data-binary "@$1"
```
Comment (mark-prins) [inline dgraph/query.sh:3]: When running `query.sh` I get `"message": "Unrecognised Content-Type.  Please use application/json for GraphQL requests"`
Changing to `application/json` works in postman, though not when using the script

File: dgraph/sample/insert.gql [hunk 2]
```diff
@@ -0,0 +1,342 @@
+mutation {
```
Comment (mark-prins) [inline dgraph/sample/insert.gql:1]: similar to the issue with `query.sh` - this doesn't work for me!
can you give me a hand to try it out?

**Comment by mark-prins** [comment]
I think delete `data.json` if the `insert.gql` works - though my preference would be to keep `data.json` as it is easier to read, and inject the content into the format required by the insert! that maybe a little OTT 🤷‍♂️

File: dgraph/sample/schema.gql [hunk 3]
```diff
@@ -1,20 +1,13 @@
-type Entity {
-  code
-  description
-  type
-  has_property
-  has_child
+type Entity @dgraph(type: "Entity") {
+    type: String! @dgraph(pred:"type") @search(by: [exact])
+    code: String @dgraph(pred:"code") @search(by: [exact])
+    description: String! @dgraph(pred:"description") @search(by: [term, exact, trigram])
```
Comment (mark-prins) [inline dgraph/sample/schema.gql:4]: when querying - I don't find the `anyoftext` very helpful for what we need. Searching for `anyoftext: "Amox"` returns no results. Using the `fulltext` is more like what we need: `regexp: "/Amox/"` will return results.

```suggestion
    description: String! @dgraph(pred:"description") @search(by: [fulltext, exact])
```

Comment (mark-prins) [reply dgraph/sample/schema.gql:4]: Yes, `anyoftext: "Amoxicillin"` returns results - that's not how mSupply is using the search atm though!

File: dgraph/query.sh (hunk 1 above)
Comment (katherine-sussol) [reply dgraph/query.sh:3]: Oops - yes I'm getting the same and did most of my testing via postman, it looks like postman/other graphQL tools do a few modifications behind the scenes before sending the payload off! How misleading - will change the insert.gql to reflect what is actually being sent to the server (required format: https://graphql.dgraph.io/api/)

File: dgraph/sample/schema.gql (hunk 3 above)
Comment (katherine-sussol) [reply dgraph/sample/schema.gql:4]: Makes sense, seems like mSupply is doing more of a partial word match as entered in the UI? Sounds like a more useful search experience too 👍

File: dgraph/query.sh (hunk 1 above)
Comment (katherine-sussol) [reply dgraph/query.sh]: Fixed (and tested via script!) with e4c8c29996f4e8c21348da4fe019cf4232a94de2 (though insertMutation.json looks a little ugly now)

File: dgraph/sample/insert.gql (hunk 2 above)
Comment (katherine-sussol) [reply dgraph/sample/insert.gql]: 🤦‍♀️  wasn't just you! It should be fixed with e4c8c29996f4e8c21348da4fe019cf4232a94de2

File: dgraph/query.sh (hunk 1 above)
Comment (katherine-sussol) [reply dgraph/query.sh]: Not relevant at this point (but will be later) - seems like in a future official release (it's currently in beta) Dgraph will support Content-Type application/graphql and treat body as graphql request: https://github.com/dgraph-io/dgraph/pull/5125 🥳

**Pull request merged**